cli/tools/lint/rules/mod.rs

```diff
@@ -0,0 +1,296 @@
+
+  fn help_docs_url(&self) -> Cow<'static, str>;
+
+  fn into_base(self: Box<Self>) -> Box<dyn LintRule>;
+}
+
+pub enum FileOrPackageLintRule {
+  File(Box<dyn LintRule>),
+  Package(Box<dyn PackageLintRule>),
+}
+
+#[derive(Debug)]
+enum CliLintRuleKind {
+  DenoLint(Box<dyn LintRule>),
+  Extended(Box<dyn ExtendedLintRule>),
+  Package(Box<dyn PackageLintRule>),
+}
+
+#[derive(Debug)]
+pub struct CliLintRule(CliLintRuleKind);
+
+impl CliLintRule {
+  pub fn code(&self) -> &'static str {
+    use CliLintRuleKind::*;
+    match &self.0 {
+      DenoLint(rule) => rule.code(),
+      Extended(rule) => rule.code(),
+      Package(rule) => rule.code(),
+    }
+  }
+
+  pub fn tags(&self) -> &'static [&'static str] {
+    use CliLintRuleKind::*;
+    match &self.0 {
+      DenoLint(rule) => rule.tags(),
+      Extended(rule) => rule.tags(),
+      Package(rule) => rule.tags(),
+    }
+  }
+
+  pub fn docs(&self) -> &'static str {
+    use CliLintRuleKind::*;
+    match &self.0 {
+      DenoLint(rule) => rule.docs(),
+      Extended(rule) => rule.docs(),
+      Package(rule) => rule.docs(),
+    }
+  }
+
+  pub fn help_docs_url(&self) -> Cow<'static, str> {
+    use CliLintRuleKind::*;
+    match &self.0 {
+      DenoLint(rule) => {
+        Cow::Owned(format!("https://lint.deno.land/rules/{}", rule.code()))
+      }
+      Extended(rule) => rule.help_docs_url(),
+      Package(rule) => rule.help_docs_url(),
+    }
+  }
+
+  pub fn supports_incremental_cache(&self) -> bool {
+    use CliLintRuleKind::*;
+    match &self.0 {
+      DenoLint(_) => true,
+      Extended(rule) => rule.supports_incremental_cache(),
+      // graph rules don't go through the incremental cache, so allow it
+      Package(_) => true,
+    }
+  }
+
+  pub fn into_file_or_pkg_rule(self) -> FileOrPackageLintRule {
+    use CliLintRuleKind::*;
+    match self.0 {
+      DenoLint(rule) => FileOrPackageLintRule::File(rule),
+      Extended(rule) => FileOrPackageLintRule::File(rule.into_base()),
+      Package(rule) => FileOrPackageLintRule::Package(rule),
+    }
+  }
+}
+
+#[derive(Debug)]
+pub struct ConfiguredRules {
+  pub all_rule_codes: HashSet<&'static str>,
+  pub rules: Vec<CliLintRule>,
+}
+
+impl ConfiguredRules {
+  pub fn incremental_cache_state(&self) -> Option<impl std::hash::Hash> {
+    if self.rules.iter().any(|r| !r.supports_incremental_cache()) {
+      return None;
+    }
+
+    // use a hash of the rule names in order to bust the cache
+    let mut codes = self.rules.iter().map(|r| r.code()).collect::<Vec<_>>();
+    // ensure this is stable by sorting it
+    codes.sort_unstable();
+    Some(codes)
+  }
+}
+
+pub struct LintRuleProvider {
+  sloppy_imports_resolver: Option<Arc<SloppyImportsResolver>>,
+  workspace_resolver: Option<Arc<WorkspaceResolver>>,
+}
+
+impl LintRuleProvider {
+  pub fn new(
+    sloppy_imports_resolver: Option<Arc<SloppyImportsResolver>>,
+    workspace_resolver: Option<Arc<WorkspaceResolver>>,
+  ) -> Self {
```

Comment:
Consider doing options object instead

Reply:
The "rule" I've been following is an options object for configuration settings and then just regular parameters for "service structs". That also makes it easier to implement `Default` for the options object (though in this case that wouldn't be a problem because both parameters happen to be optional). I don't think introducting explanatory options is that helpful in this scenario.